### DISCUSSION TITLE
Print a PDF on a remote printer

### DISCUSSION THREAD

**Discussion by mpapini64** [nextgenhealthcare/connect#5563] [Q&A] [answered]
Hi all, I have the following need.
I have a PDF created with a Destination "Document Writer" starting from an HTML template, using an XML as data source.
The PDF is created correctly, but I need to print it on a network printer and not on a local printer.
The printer is however registered on the server where Mirth runs with his name.
I found this snipet:
```
process = java.lang.Runtime.getRuntime().exec("java -jar \"C:\\Program Files\\Mirth Connect\\extensions\\doc\\lib\\pdfbox-app-2.0.1.jar \" PrintPDF -silentPrint -printerName " + $('Printer') + " \"" + pdfPath+pdfFilename + "\"");
// If Returncode equals Zero (== 0), printing was successfull
if ( process.waitFor() == 0 )
var PrintPdfSuccess = true;
else
var PrintPdfSuccess = false;
```

But it seems to only work with local printers!
Is it possible to call the acrobat reader from the command line by indicating to print the pdf on the "PrinterRemoteStaz20" queue?

**Comment by siddharthsharma9258** [comment]
Even before you can print on the network printer. You need to check whether the printer is reachable or not. 
And how are you sure that local printer is working? Did you get a physical printout of your pdf?

**Reply by mpapini64** [reply]
On the server I have no local printers, the printers are only on the network, all registered on the server, but all network, each with its own name.
Yes of course, it would be interesting to be able to test the actual availability of the printer (is there a way?), but basically they are always available as they are printers near the acceptance station, where a CUP prints its receipts and we, who are a LIS, we have to print a sort of slip containing the list of exams and the report delivery date.
We agree that it is an absurd system and it makes me vomit, but the CUP is not able to print anything coming from the LIS and therefore we have to print it ourselves, remotely.

**Reply by pacmano1** [reply]
https://forums.mirthproject.io/forum/mirth-connect/support/15007-howto-printing-from-mirth-and-a-question?p=180144#post180144

Just an alternate idea I posted.

**Reply by mpapini64** [reply]
Unfortunately my server is under windows!
Anyway thanks for the answer!

**Comment by odoodo** [comment, marked as answer]
**NetworkPrinting**
A small Java library that allows to print documents of type
* PDF
* HTML
* Text

from within Mirth on the fly to a network printer w/o having a driver installed locally.

**Functionality**
Print a pdf document represented by a byte array:
```js
Packages.lu.hrs.mirth.NetworkPrinting.printPdf(documentToPrintAsByteArray, printerIpOrName);`
```
Print a base64-encoded document:
```js
Packages.lu.hrs.mirth.NetworkPrinting.printPdfBase64(base64EncodedDocumentToPrint, printerIpOrName);
```
Print an HTML page:
```js
Packages.lu.hrs.mirth.NetworkPrinting.printHtml(htmlCode, printerIpOrName);
```
Print text:
```js
Packages.lu.hrs.mirth.NetworkPrinting.printText(textToPrint, printerIpOrName, encoding);
```
**Installation**
1. unpack the [attached folder](https://github.com/nextgenhealthcare/connect/files/10186546/NetworkPrinting.zip)
2. place the whole folder in the custom-lib folder of your Mirth installation
3. reload resources (assure that _Include all Subdirectories_-checkbox is checked)

**Source code**
is included in the jar and also [added to this post](https://github.com/nextgenhealthcare/connect/files/10186572/NetworkPrinting-Sources.zip) for convenience

**Reply by mpapini64** [reply]
Your solution is great! Thank You very much!!!
